common/scala/src/main/scala/whisk/core/containerpool/ContainerClient.scala

```diff
@@ -0,0 +1,218 @@
+ * @param queueSize once all connections are used, how big of queue to allow for additional requests
+ * @param retryInterval duration between retries for TCP connection errors
+ */
+protected class PoolingContainerClient(
+  hostname: String,
+  port: Int,
+  timeout: FiniteDuration,
+  maxResponse: ByteSize,
+  queueSize: Int,
+  retryInterval: FiniteDuration = 100.milliseconds)(implicit logging: Logging, as: ActorSystem)
+    extends PoolingRestClient("http", hostname, port, queueSize, timeout = Some(timeout))
+    with ContainerClient
+    with AutoCloseable {
+
+  def close() = Await.result(shutdown(), 30.seconds)
+
+  /**
+   * Posts to hostname/endpoint the given JSON object.
+   * Waits up to timeout before aborting on a good connection.
+   * If the endpoint is not ready, retry up to timeout.
+   * Every retry reduces the available timeout so that this method should not
+   * wait longer than the total timeout (within a small slack allowance).
+   *
+   * @param endpoint the path the api call relative to hostname
+   * @param body the JSON value to post (this is usually a JSON objecT)
+   * @param retry whether or not to retry on connection failure
+   * @return Left(Error Message) or Right(Status Code, Response as UTF-8 String)
+   */
+  def post(endpoint: String, body: JsValue, retry: Boolean)(
+    implicit tid: TransactionId): Future[Either[ContainerHttpError, ContainerResponse]] = {
+
+    //create the request
+    val req = Marshal(body).to[MessageEntity].map { b =>
+      //DO NOT reuse the connection (in case of paused containers)
+      //For details on Connection: Close handling, see:
+      // - https://doc.akka.io/docs/akka-http/current/common/http-model.html#http-headers
+      // - http://github.com/akka/akka-http/tree/v10.1.3/akka-http-core/src/test/scala/akka/http/impl/engine/rendering/ResponseRendererSpec.scala#L470-L571
+      HttpRequest(HttpMethods.POST, endpoint, entity = b).withHeaders(Connection("close"))
+    }
+
+    retryingRequest(req, timeout, retry)
+      .flatMap(r => {
+        if (r._2 > 0) {
+          logging.info(this, s"completed after ${r._2} retries")
+        }
+        val response = r._1
+        response.entity.contentLengthOption match {
+          case Some(contentLength) if response.status != StatusCodes.NoContent =>
+            if (contentLength <= maxResponse.toBytes) {
+              Unmarshal(response.entity.withSizeLimit(maxResponse.toBytes)).to[String].map { o =>
+                Right(ContainerResponse(response.status.intValue, o, None))
+              }
+            } else {
+              truncated(response.entity.dataBytes).map { s =>
+                Right(ContainerResponse(response.status.intValue, s, Some(contentLength.B, maxResponse)))
+              }
+            }
+          case _ =>
+            //handle missing Content-Length as NoResponseReceived
+            //also handle 204 as NoResponseReceived, for parity with HttpUtils client
```

Comment:
I think we still need to drain the entity, like:

```scala
response.discardEntityBytes().future.map(_ => Left(NoResponseReceived())
```

since this case is also reached by an unknown Content-Length.